### ISSUE TITLE
"No preview available" on all pixelfed.social on federated instance

### ISSUE THREAD

**Issue by nehemiagurl** [pixelfed/pixelfed#4853]
I'm on anar.chi.st, which runs 0.11.9. every post from pixelfed.social appears as "no preview available".
for example, https://pixelfed.social/p/TLENick/649164566507970710 renders as:
<img width="713" alt="Screenshot 2024-01-06 at 16 35 38" src="https://github.com/pixelfed/pixelfed/assets/109424320/8de8440d-0cb1-4903-abe1-3819ba25818b">
compare with https://pixelfed.fr/p/raph/649009863429882318 which renders normally:
<img width="685" alt="Screenshot 2024-01-06 at 16 39 47" src="https://github.com/pixelfed/pixelfed/assets/109424320/7f87a797-51cd-4ae3-9575-a1dfef6ad8c6">

this appears to only be a pixelsoc thing. mastodon posts render normally, other instances that also run .9 also render normally.

**Comment by dansup** [comment]
I can't reproduce this on pixelfed.art or piconic.co, please check with your admins and ensure they are running the latest version!

**Comment by nehemiagurl** [comment]
as I mentioned in the initial report, the instance does run the latest version, 0.11.9. @dansup 
<img width="368" alt="Screenshot 2024-01-07 at 10 53 27" src="https://github.com/pixelfed/pixelfed/assets/109424320/125f3af7-b70f-4f1d-aaee-32d19ad1f7fd">

**Comment by nehemiagurl** [comment]
@dansup this is definitely still a thing, and as mentioned the instance _is_ running the latest version.

**Comment by nehemiagurl** [comment]
also, please note that this issue is not reproducible on Safari or on Firefox mobile.

**Comment by nehemiagurl** [comment]
still happening on .11.11 @dansup. any chance of reopening this issue?

**Comment by skycorrigan** [comment]
Add me to the list of users with "no preview available" problems on version 0.11.11.

**Comment by jesseplusplus** [comment]
I am also seeing this issue on pixelfed.social loading a post from frequency.app. I noticed that the request for the image from the original server was failing because the url is quite long (it's a signed url) and was getting truncated (see screenshot below). 

Can others confirm if this is also happening for them?

<img width="1425" alt="pixelfed-no-preview" src="https://github.com/pixelfed/pixelfed/assets/382669/81898c5d-878f-4fc9-9b90-4aca536652a6">

**Comment by nehemiagurl** [comment]
@jesseplusplus good insight! I'm getting blocks from pxscdn.com when viewing via anar.chi.st, as well as when going to pixelfed.social directly:
<img width="767" alt="Screenshot 2024-03-13 at 13 44 01" src="https://github.com/pixelfed/pixelfed/assets/109424320/9741547b-b4f3-4e0f-a43b-9edaaff18a24">
<img width="769" alt="Screenshot 2024-03-13 at 13 45 20" src="https://github.com/pixelfed/pixelfed/assets/109424320/905759f2-754d-4576-9b87-6a6d1eb0aaca">

**Comment by jesseplusplus** [comment]
Okay, I think I've tracked down why this is happening:
  
**tl;dr**: this is due to strings stored in the `media_path` column of the `media` table being truncated to the default length of 191 characters so that they will fit in the index. This restriction is only necessary for older versions of InnoDB (MySQL/MariaDB) databases that haven't had their row format changed to new the default and the large prefix flag enabled (which is now the default config); the index size restriction is not an issue on Postgres. A migration to allow enlarging this column would fix this.  
  
---
  
Details:
  
The urls my mastodon instance was sending to pixelfed.social were being truncated at 191 characters. For privacy on followers-only media I'm using s3 signed urls, which are quite long – the signature itself is 268 characters.  
  
When a remote status with media attachments is ingested into pixelfed, it creates an entry in the database's `media` table and stores the url string in the `media_path` column. This `media_path` column [was created with the default string length](https://github.com/pixelfed/pixelfed/blob/8aae92d75b4881e5cd6a4e2c83ca52046e5b3d66/database/migrations/2018_04_16_011918_create_media_table.php#L21). The default string length in laravel [is 255 characters](https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Schema/Blueprint.php#L776-L778), but in older versions it was 191 for MySQL/MariaDB if using utf8mb4 encoding (which is the default laravel encoding) and `COMPACT` row format. This is due to the following constraint in older versions of InnoDB as mentioned [here](https://laracasts.com/discuss/channels/general-discussion/migrations-configure-a-default-string-length-utf8mb4?page=1&replyId=143701):  
  
> InnoDB has a maximum index length of 767 bytes, so for [...] utf8mb4 columns, you can index a maximum of [...] 191 characters.  
  
The `media` table has a [unique index on the combinations of `status_id` and `media_path`](https://github.com/pixelfed/pixelfed/blob/8aae92d75b4881e5cd6a4e2c83ca52046e5b3d66/database/migrations/2018_04_16_011918_create_media_table.php#L31), so this length constraint on the `media_path` column was necessary for the functioning of that index on earlier InnoDB versions, but this is no longer the case.
  
Some further explanation from the MySQL and MariaDB docs:  

- [MySQL](https://dev.mysql.com/doc/refman/5.7/en/innodb-limits.html):
	    
  > The index key prefix length limit is 767 bytes for  `InnoDB` tables that use the [REDUNDANT](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_redundant_row_format) or [COMPACT](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_compact_row_format) row format. For example, you might hit this limit with a [column prefix](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_column_prefix) index of more than 255 characters on a `TEXT` or `VARCHAR` column, assuming a `utf8mb3` character set and the maximum of 3 bytes for each character.  
	    
  > The limits that apply to index key prefixes also apply to full-column index keys.  

  - From the [`COMPACT` row format docs:](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_compact_row_format)
		    
    > The default `InnoDB` **row format** for InnoDB tables from MySQL 5.0.3 to MySQL 5.7.8.   
		    
    > As of MySQL 5.7.9, the default row format is defined by the [`innodb_default_row_format`](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_default_row_format) configuration option, which has a default setting of **DYNAMIC**.  

- [MariaDB](https://mariadb.com/kb/en/innodb-row-formats-overview/#row-formats):
	    
  > `DYNAMIC` is the default row format.  
	    
  > Default row format in [MariaDB 10.2.1](https://mariadb.com/kb/en/mariadb-1021-release-notes/) and earlier `COMPACT`.  

In these older versions of the databases, you *can* have longer strings and be able to index them properly with two configuration changes, by:

- [Changing the default row format](https://dev.mysql.com/doc/refman/5.7/en/innodb-row-format.html#innodb-row-format-dynamic)
  > The `DYNAMIC` row format offers the same storage characteristics as the `COMPACT` row format but adds enhanced storage capabilities for long variable-length columns and supports large index key prefixes.  

- [and setting the`innodb_large_prefix` flag:](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix)
	    
  > When this option is enabled, index key prefixes longer than 767 bytes (up to 3072 bytes) are allowed for `InnoDB` tables that use [DYNAMIC](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_dynamic_row_format) or [COMPRESSED](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_compressed_row_format) row format. See [Section 14.23, “InnoDB Limits”](https://dev.mysql.com/doc/refman/5.7/en/innodb-limits.html) for maximums associated with index key prefixes under various settings.  
	    
  > For tables that use [REDUNDANT](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_redundant_row_format) or [COMPACT](https://dev.mysql.com/doc/refman/5.7/en/glossary.html#glos_compact_row_format) row format, this option does not affect the permitted index key prefix length.  
	    
  > [`innodb_large_prefix`](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix) is enabled by default in MySQL 5.7.  

---

It looks like fixing this would require:

* a config change for MySQL <5.7. to enable `innodb_large_prefix`
* a migration to alter the `media` table to `ROW_FORMAT=DYNAMIC`
* a migration to change the `media_path` to a longer length (up to 768 characters).
  
@dansup does this sound reasonable? Is there any reason I'm missing that you wouldn't want to change these configurations and sizes?

**Comment by nehemiagurl** [comment]
very strangely, it seems that some posts slip through. here are three posts in a row from the same person - two are visible, one is not. even when I see them individually and not as part of a feed.

https://github.com/pixelfed/pixelfed/assets/109424320/3a123426-d131-4a68-b190-7a422791d79f

and of course, if I try to open them in pixelfed.social, I still get blocked...

the posts are:
https://pixelfed.social/p/Wizard_Wabash/683022786426406171
https://pixelfed.social/p/Wizard_Wabash/683012026288904119
https://pixelfed.social/p/Wizard_Wabash/683007832250137619

and their view links from anar.chi.st are:
https://anar.chi.st/i/web/post/683022795095765452
https://anar.chi.st/i/web/post/683012045018726595
https://anar.chi.st/i/web/post/683007853227696213

**Comment by MarkoKaartinen** [comment]
I've gotten reports from my users also that some posts shows no preview available.
For example this post: https://pixelfed.social/p/Muiqthekhajiit/791592047868797533

Through pixl.fi it shows no preview and throudh my other pixelfed instance it shows it.
Console shows 
(failed)net::ERR_BLOCKED_BY_ORB
which is bit odd since it shows 404 when opening via inspector and when opening. the single image it redirects to pxscdn 

<img width="486" alt="Image" src="https://github.com/user-attachments/assets/abd7d49d-cd75-4394-bfe5-dad5afe2e8ee" />

Any tips/tricks to get these working again?
